[JENKINS-71971][JEP-237] FIPS-140 compliant version of HudsonPrivateSecurityRealm

divyasivasamy opened the pull request:
[JENKINS-71971](https://issues.jenkins.io/browse/JENKINS-71971) Allow the use of `HudsonPrivateSecurityRealm` with a FIPS compliant password algorithm.

Prior to this change the internal security realm used bcrypt (which uses the blowfish algorithm) to hash passwords.  Blowfish is not a FIPS compliant algorithm and as this was used in the initial install it meant a clean install setup via the wizard could never be FIPS-140 compliant.  
This addresses that by Implementing an alternative (enabled by a `SystemProperty` that uses `PBKDF2withHmacSHA512`.

see also [JEP-237](https://github.com/jenkinsci/jep/tree/master/jep/237)

### Testing done

To validate the FIPS Mode and create hash of the user password and store it.
to validate a given password during the authentication process against stored hash of the password  

### Proposed changelog entries

* Developer: Optionally support a FIPS-140 compliant algorithm in the Jenkins' own user database.

### Proposed upgrade guidelines

N/A

divyasivasamy committed:
Password Hashing changes in HudsonPrivateSecurityRealm

divyasivasamy committed:
Password Hashing -PBKDF2 Implementation

divyasivasamy committed:
reverted docs

divyasivasamy committed:
isHashValid Changes

divyasivasamy committed:
isHashValid Changes

divyasivasamy committed:
commented logger

divyasivasamy committed:
commented logger

divyasivasamy committed:
modified test cases

divyasivasamy committed:
modified test cases

divyasivasamy committed:
modified test cases

divyasivasamy committed:
Merge branch 'master' into HudsonPrivateSecurityRealmImplementation

divyasivasamy committed:
resolved Util conflicts

divyasivasamy committed:
resolved Util conflicts

divyasivasamy committed:
resolved Util conflicts

divyasivasamy committed:
modified test cases

jtnord commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -925,49 +933,172 @@ public boolean isHashValid(String hash) {
+        private static byte[] generateSalt() {
+            SecureRandom random = new SecureRandom();
+            byte[] salt = new byte[16];
+            random.nextBytes(salt);
```
should we be using `generateSeed(salt)`  (one for the security team)

jtnord replied:
we are fine with `nextBytes` here, we are not seeding a new RNG

jtnord commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -925,49 +933,172 @@ public boolean isHashValid(String hash) {
+            return diff == 0;
+        }
+
+        private static byte[] fromHex(String storedPassword)  {
```
`hudson.util.`[`fromHexString`](https://javadoc.jenkins.io/hudson/Util.html#fromHexString(java.lang.String))​ ?

jtnord replied:
remote this code then?

jtnord commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -925,49 +933,172 @@ public boolean isHashValid(String hash) {
+     * Magic header used to detect if a password is hashed.
+     */
+
+    public static String getPasswordHeader() {
+        return Util.isFipsMode() ? PBKDF2 : JBCRYPT;
+    }
```
I think it is valid to be able to go from FIPS to non FIPS without loosing all your users?  (but the reverse is not valid).

divyasivasamy replied:
It wouldn't work if we change the mode .

jtnord commented on test/src/test/java/hudson/security/HudsonPrivateSecurityRealmTest.java:
```diff
@@ -498,11 +519,69 @@ public void controlCharacterAreNoMoreValid_CustomRegex() throws Exception {
     }
 
+    @Test
+    public void passwordHashWithPBKDF2() {
```
there is a Jenkins started and running for all of these tests - but it does not appear that you need one.
if that is the case the test methods can be annotated with [`@WithoutJenkins`](https://javadoc.jenkins.io/component/jenkins-test-harness/org/jvnet/hudson/test/WithoutJenkins.html)

divyasivasamy replied:
updated test method with @WithoutJenkins

jtnord commented on test/src/test/java/hudson/security/HudsonPrivateSecurityRealmTest.java:
```diff
@@ -498,11 +519,69 @@ public void controlCharacterAreNoMoreValid_CustomRegex() throws Exception {
+        pbkdf2PasswordEncoder.encode("password");
+    }
+
+    @Test
+    public void passwordHashWithInvalidKeySpec() throws InvalidKeySpecException {
+        HudsonPrivateSecurityRealm.PBKDF2PasswordEncoder pbkdf2PasswordEncoder = new HudsonPrivateSecurityRealm.PBKDF2PasswordEncoder();
+        mockStatic(HudsonPrivateSecurityRealm.class);
+        SecretKeyFactory secretKeyFactory  = mock(SecretKeyFactory.class);
+        when(secretKeyFactory.generateSecret(any())).thenThrow(InvalidKeySpecException.class);
+        pbkdf2PasswordEncoder.encode("password");
+    }
```
shouldn't this be expected to throw an exception given the `when` clause and if so isn't this going to be a failing test?

jtnord replied:
this is bad - `pbkdf2PasswordEncoder.encode()` is then returning as if the password it has encoded is valid - smells like a security issue.

divyasivasamy replied:
modified test cases, throws exception when key spec is not valid please verify

divyasivasamy commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -925,49 +933,172 @@ public boolean isHashValid(String hash) {
         public boolean matches(CharSequence rawPassword, String encPass) {
             if (isPasswordHashed(encPass)) {
-                return JBCRYPT_ENCODER.matches(rawPassword, encPass.substring(JBCRYPT_HEADER.length()));
+                return PASSWORD_HASH_ENCODER.matches(rawPassword, encPass.substring(getPasswordHeader().length()));
```
It wouldn't work if we change the mode .

Have implemented these changes as per the DACI 
(Decision 1: Scope of change - Option 3: -
 Its captured in additional details as - **it is not possible to change FIPS mode after installation**

![image](https://github.com/jenkinsci/jenkins/assets/130471287/0cf03528-8960-4bf5-b2d7-5793bbd5e4e4)

divyasivasamy committed:
fixed review comments

divyasivasamy commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -1032,5 +1163,6 @@ public void destroy() {
     };
 
-    private static final Logger LOGGER = Logger.getLogger(HudsonPrivateSecurityRealm.class.getName());
+     static final Logger LOGGER = Logger.getLogger(HudsonPrivateSecurityRealm.class.getName());
```
its reverted and added private

divyasivasamy committed:
fixed review comments

divyasivasamy committed:
fixed review comments

jtnord commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -925,34 +933,157 @@ public boolean isHashValid(String hash) {
+            try {
+                secretKeyFactory = SecretKeyFactory.getInstance(PBKDF2_ALGORITHM);
+                hash = secretKeyFactory.generateSecret(spec).getEncoded();
+            } catch (NoSuchAlgorithmException e) {
+                LOGGER.log(WARNING, "No such algorithm found for encode",  e);
+            } catch (InvalidKeySpecException e) {
+                LOGGER.log(WARNING, "Invalid key spec exception",  e);
+            }
```
so this returns an empty (zero length) byte array if we can not get the algotithm or the keysepc is invalid?

There are 2 things here

1. this is a silent failure as far as a user attempting to login - things could just fail with no indication things went wrong.
2. how does this impact the security - are we not going to take *any* supplied password and encode it to the same empty zero lenght byte array which means you have no security?  this would seem like an exceptionally bad security issue if true.

divyasivasamy replied:
modified to throw exception

jtnord commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -925,49 +933,172 @@ public boolean isHashValid(String hash) {
+            return hash;
+        }
+
+        private static String toHex(byte[] secretKey)  {
```
if this code is no longer user then it should be removed, if it is used the previous comment would be incorrect.

divyasivasamy replied:
Its Removed

jtnord commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -925,34 +933,157 @@ public boolean isHashValid(String hash) {
+
+            byte[] generatedHashValue = generateSecretKey(spec);
+
+            int diff = hash.length ^ generatedHashValue.length;
+            for (int i = ZERO; i < hash.length && i < generatedHashValue.length; i++)
+            {
+                diff |= hash[i] ^ generatedHashValue[i];
+            }
```
compare security tokens (byte[]) with [`MessageDigest.isEqual(byte[], byte[])`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/security/MessageDigest.html#isEqual(byte%5B%5D,byte%5B%5D)) rather than writing your own.

(will need an import)
```suggestion
            return MessageDigest.isEqual(hash, generatedHashValue);
```

divyasivasamy replied:
updated the same

jtnord commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -925,49 +933,172 @@ public boolean isHashValid(String hash) {
                 return false;
             }
-            return password.startsWith(JBCRYPT_HEADER) && JBCRYPT_ENCODER.isHashValid(password.substring(JBCRYPT_HEADER.length()));
+                return password.startsWith(getPasswordHeader()) && PASSWORD_HASH_ENCODER.isHashValid(password.substring(getPasswordHeader().length()));
```
the indentation is still wrong though.

divyasivasamy replied:
Fixed Indentation

jtnord commented on core/src/main/java/hudson/security/PasswordHashEncoder.java:
```diff
@@ -0,0 +1,7 @@
+
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+public interface PasswordHashEncoder extends PasswordEncoder {
```
does the interface need to be `public`?  can we keep it package scoped so that it does not leak and can be more easily changed later if needed

divyasivasamy replied:
yes changed

rsandell commented on core/src/main/java/hudson/Util.java:
```diff
@@ -1952,4 +1952,17 @@ public static String getHexOfSHA256DigestOf(byte[] input) throws IOException {
+    public static boolean FIPS_MODE = SystemProperties.getBoolean(Util.class.getName() + ".FIPS_MODE");
+
+    public static boolean isFipsMode() {
+        LOGGER.info("FIPS mode : " + FIPS_MODE);
```
It looks like this method is going to be called quite often, so spamming the system log with this on `INFO` level doesn't look good.

divyasivasamy replied:
modified it to FINE level

divyasivasamy committed:
fixed review comments

divyasivasamy committed:
fixed review comments

olamy commented:
looks to be related to https://github.com/jenkinsci/jep/pull/398
shouldn't we have a `fips` label for triage/search purpose? (not sure about the exact label name though :) )

jtnord commented:
> looks to be related to [jenkinsci/jep#398](https://github.com/jenkinsci/jep/pull/398) shouldn't we have a `fips` label for triage/search purpose? (not sure about the exact label name though :) )

Good point - I was wondering something similar earlier if we should have a `needs-fips-review` similar to `needs-security-review` for things that look like they may impact FIPS compliance (for the better or worse), but for a different reason.

Shall we move this conversation to the [dev mailing list](https://groups.google.com/forum/#!forum/jenkinsci-dev)?  (or perhaps this should be part of https://github.com/jenkinsci/jep/pull/398)

jtnord committed:
Apply suggestions from code review

jtnord committed:
checkstyle

jtnord committed:
introduce specific class for FIPS

jtnord committed:
Merge remote-tracking branch 'origin/master' into HudsonPrivateSecurityRealmImplementation

jtnord committed:
misc code cleanups

jtnord committed:
revert changes to HudsonPrivateSecurityRealmTest

jtnord committed:
code fixups and moving tests

jtnord committed:
code fixups and moving tests

jtnord committed:
make methods and fields private, and add message to exception

jtnord committed:
SecureRandom s -> SecureRandoms

jtnord committed:
fixup IT after increasing the number of hasing rounds

jtnord commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -925,34 +933,124 @@ public boolean isHashValid(String hash) {
+
+        @SuppressFBWarnings(value = "DMI_RANDOM_USED_ONLY_ONCE", justification = "SpotBugs you are drunk! https://github.com/spotbugs/spotbugs/issues/2128")
+        private synchronized byte[] generateSalt() {
+            // synchronized for lazy initialization but also not all SecureRandoms are thread safe.
```
not all are providers are thread safe but all SecureRandoms are thread safe

jtnord committed:
move SecureRandom to double-checked-locking as SecureRandom itself is thread safe

jtnord commented on core/src/main/java/jenkins/security/FIPS140.java:
```diff
@@ -0,0 +1,60 @@
+
+    private static final Logger LOGGER = Logger.getLogger(FIPS140.class.getName());
+
+    private static /* almost final */ boolean FIPS_COMPLIANCE_MODE = SystemProperties.getBoolean(FIPS140.class.getName() + ".COMPLIANCE");
```
```suggestion
    private static final boolean FIPS_COMPLIANCE_MODE = SystemProperties.getBoolean(FIPS140.class.getName() + ".COMPLIANCE");
```

jtnord committed:
make FIPS_COMPLIANCE_MODE  final

changing this at run time makes no sense as decisions will already have been made to configure things differently.

yaroslavafenkin reviewed (commented):
Seems fine to me security wise. Please let others review too though, I know @Kevin-CB was looking into this.

Kevin-CB reviewed (approved):
I tested this PR locally and it looks fine to me. My only concern was about a possible switch from a non-FIPS-compliant instance to a FIPS-compliant one and vice versa, but I suppose that's not within the scope, as per https://github.com/jenkinsci/jep/blob/master/jep/237/README.adoc#backwards-compatibility.

daniel-beck commented on core/src/main/java/hudson/security/PasswordHashEncoder.java:
Missing license header

daniel-beck commented on core/src/main/java/jenkins/security/FIPS140.java:
```diff
@@ -0,0 +1,60 @@
+ * Utilities to help code change behaviour when it is desired to run in a FIPS-140 enabled environment.
+ * The environment (host, JVM and servlet container), must be suitably configured which is outside the scope of the Jenkins project.
+ * @see <a href="https://csrc.nist.gov/pubs/fips/140-2/upd2/final">FIPS-140-2</a>
+ * @see <a href="https://github.com/jenkinsci/jep/pull/398/files">JEP-XXXX</a>
```
```suggestion
 * @see <a href="https://github.com/jenkinsci/jep/blob/master/jep/237/README.adoc">JEP-237</a>
```

jtnord replied:
this file will be merged as part of https://github.com/jenkinsci/jenkins/pull/8482 where it is updated.

daniel-beck commented on core/src/test/java/hudson/security/HudsonPrivateSecurityRealmTest.java:
```diff
@@ -0,0 +1,134 @@
+    private static final String PBKDF2_HMAC_SHA512_ENCODED_PASSWORD =
+            "$HMACSHA512:210000:30f9e0a5470a8bc67f128ca1aae25dd4$88abaca4f442caeff0096ec0f75df2d77cc31a956c564133232f4d2532a72c8d4380a718d5b2a3dccab9e752027eeadd8f9f2c0c624505531bf3a57ec7d08aad";
+
+    //@Test
```
Purpose?

jtnord replied:
incase people want to check the complexity of how long this takes (ie is the number of itterations we use correct for the state of CPUs).

poor mans benchmark that is not excersized but that someone can run pretty quickly if they wanted to check.

I *could* make this a test by asserting the test took > X to run - but that seemed like a test that would just waste time.

jtnord replied:
https://github.com/jenkinsci/jenkins/blob/f37bcc31de0a041a2f1782fa3bf253701d04ebcd/core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java#L950 FTR

jtnord replied:
resolved in 15e6c88be791c73f1e4a4b4afeabac0473ec7ce9

daniel-beck replied:
> FTR

FWIW I plan to submit https://github.com/daniel-beck/jenkins/commit/93718b019ec17c859fbef4fe009a671875bf78d1 once this PR is in (unnecessary merge conflict)

daniel-beck commented on core/src/test/java/hudson/security/HudsonPrivateSecurityRealmTest.java:
```diff
@@ -0,0 +1,134 @@
+        System.out.println("PBKDF2 took " + d.toSeconds() + "s");
+    }
+
+    //@Test
```
Purpose?

jtnord replied:
for comparison with PBKDF - so that someone should they want to check the difference between the current bcrypt and the new PBKDF implementation.

daniel-beck replied:
This should be documented in the code, or the code should be removed. Right now it's undocumented dead code in the form of a test without `@Test` and assertions.

jtnord replied:
15e6c88be791c73f1e4a4b4afeabac0473ec7ce9

daniel-beck commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -524,10 +530,11 @@ public User createAccount(String userName, String password) throws IOException {
+     * @see #getPasswordHeader()
      */
     public User createAccountWithHashedPassword(String userName, String hashedPassword) throws IOException {
         if (!PASSWORD_ENCODER.isPasswordHashed(hashedPassword)) {
```
Next line could probably be enhanced to identify a wrong hash format and if so, state that.

jtnord replied:
f37bcc31de0a041a2f1782fa3bf253701d04ebcd

daniel-beck reviewed (commented):
To make it easier to identify bad configurations, it would be useful to identify when a stored hashed password does not match the current value of `FIPS140#useCompliantAlgorithms` and emit warnings to the log with an explanation when that happens.

(Providing diagnostic information in case of bad system configuration does not go against https://github.com/jenkinsci/jep/blob/master/jep/237/README.adoc#backwards-compatibility)

jtnord commented on core/src/main/java/hudson/security/PasswordHashEncoder.java:
```diff
@@ -0,0 +1,7 @@
+package hudson.security;
```
```suggestion
/*
 * The MIT License
 *
 * Copyright (c) 2023, Cloudbees, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package hudson.security;
```

jtnord committed:
Add (c) header

jtnord committed:
keep the checkstyle overlord happy

jtnord committed:
update exception message depending on the specific error issue

jtnord commented:
adding `security-approved` as both @daniel-beck and @Kevin-CB have reviewed and left no security related issues.

daniel-beck commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -925,34 +941,131 @@ public boolean isHashValid(String hash) {
-                return JBCRYPT_ENCODER.matches(rawPassword, encPass.substring(JBCRYPT_HEADER.length()));
+                return PASSWORD_HASH_ENCODER.matches(rawPassword, encPass.substring(getPasswordHeader().length()));
             } else {
                 return false;
```
This is probably a second location where a check for the correct hashing algorithm (and emitting diagnostic information to the log if incorrect) could be useful.

Basically, attempting to log in while the stored password hash's format does not match the running Jenkins's expectations should result in corresponding log messages.

(This applies in both directions; someone might attempt to introduce FIPS mode late, or lose the FIPS system property during upgrade/reconfiguration).

(Ideally with test.)

jtnord commented on core/src/test/java/hudson/security/HudsonPrivateSecurityRealmTest.java:
```diff
@@ -0,0 +1,134 @@
+    private static final String PBKDF2_HMAC_SHA512_ENCODED_PASSWORD =
+            "$HMACSHA512:210000:30f9e0a5470a8bc67f128ca1aae25dd4$88abaca4f442caeff0096ec0f75df2d77cc31a956c564133232f4d2532a72c8d4380a718d5b2a3dccab9e752027eeadd8f9f2c0c624505531bf3a57ec7d08aad";
+
+    //@Test
```
```suggestion
    /* This exists so that we can easily  check the complexity of how long this takes (ie is the number of iterations we 
     * use correct for the state of CPUs).  
     * We do not want to assert that the range < x and  > y  as that would make the test flaky on overloaded
     *  or slow hardware, so this is commented out but left for ease of running locally when desired.
     */
    //@Test
```

jtnord commented on core/src/test/java/hudson/security/HudsonPrivateSecurityRealmTest.java:
```diff
@@ -0,0 +1,134 @@
+        System.out.println("PBKDF2 took " + d.toSeconds() + "s");
+    }
+
+    //@Test
```
```suggestion
    /* This exists so that we can easily  check the complexity of how long this takes (ie is the number of iterations we 
     * use correct for the state of CPUs).  
     * We do not want to assert that the range < x and  > y  as that would make the test flaky on overloaded
     *  or slow hardware, so this is commented out but left for ease of running locally when desired.
     */
    //@Test
```

jtnord committed:
Leave comment about commented out tests

jtnord commented on core/src/test/java/hudson/security/HudsonPrivateSecurityRealmTest.java:
```diff
@@ -0,0 +1,144 @@
+    /* This exists so that we can easily  check the complexity of how long this takes (ie is the number of iterations we 
+     * use correct for the state of CPUs).  
+     * We do not want to assert that the range < x and  > y  as that would make the test flaky on overloaded
+     *  or slow hardware, so this is commented out but left for ease of running locally when desired.
```
```suggestion
     * or slow hardware, so this is commented out but left for ease of running locally when desired.
```

jtnord commented on core/src/test/java/hudson/security/HudsonPrivateSecurityRealmTest.java:
```diff
@@ -0,0 +1,144 @@
+    /* This exists so that we can easily  check the complexity of how long this takes (ie is the number of iterations we 
+     * use correct for the state of CPUs).  
+     * We do not want to assert that the range < x and  > y  as that would make the test flaky on overloaded
+     *  or slow hardware, so this is commented out but left for ease of running locally when desired.
```
```suggestion
     * or slow hardware, so this is commented out but left for ease of running locally when desired.
```

jtnord committed:
remove extra space in comment

daniel-beck commented on core/src/test/java/hudson/security/HudsonPrivateSecurityRealmTest.java:
```diff
@@ -0,0 +1,144 @@
+                "not enough iterations");
+        assertFalse(encoder.isHashValid(
+                "$HMACSHA512:500000:fe899fcfcef4302ec3f0d36164efefdc$0781364eae9dac4ef1c4c3bf34c28e13965b46105fec0b6fcf4bae78e246fb5e51a1694fff19acac2dfb37b16055092644f3682c25beea9a7a286bf94e52f63b"),
+                "too many iterations");
```
Why is this a problem? (unreasonable number of iterations, maybe, but this is still the same order of magnitude).

jtnord replied:
it does not match the Regular expression that this is specifically testing - it does not matter if we are over by 1, one thousand, or 1 million.

daniel-beck replied:
Seems unnecessarily restrictive, and the test alone does not indicate why this is expected to fail. Just a nit though.

daniel-beck commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -925,34 +941,131 @@ public boolean isHashValid(String hash) {
+
+        private SecureRandom secureRandom() {
+            // lazy initialisation so that we do not block startup due to entropy
+            if (random == null) {
+                synchronized (this) {
+                    if (random == null) {
+                        random = new SecureRandom();
+                    }
+                }
+            }
```
Is this accessed frequently enough to need double-checked locking and a `volatile` field?

jtnord replied:
probably not - but this is a semi standard idiom for lazy initialization.
Once you have created 2 or more passwords this will be cheaper than synchronization and does no harm

daniel-beck reviewed (commented):
I see no major problems with this PR.

daniel-beck commented:
Spotless failed the build.

IMO Spotless means applying PR suggestions has a very small chance of being successful, so should just not be done.

jtnord committed:
remove trailing line ends to keep spotless happy

daniel-beck committed:
Merge branch 'master' into HudsonPrivateSecurityRealmImplementation

jtnord committed:
warn when password hashed with incorrect algorithm

daniel-beck commented on test/src/test/java/hudson/security/HudsonPrivateSecurityRealmTest.java:
```diff
@@ -70,15 +74,23 @@
 @For({UserSeedProperty.class, HudsonPrivateSecurityRealm.class})
 public class HudsonPrivateSecurityRealmTest {
 
+    // the PBKDF encoded for of "password" without the quotes
```
```suggestion
    // the PBKDF encoded form of "password" without the quotes
```

daniel-beck commented on test/src/test/java/hudson/security/HudsonPrivateSecurityRealmTest.java:
```diff
@@ -717,6 +730,40 @@ public void changingPassword_withSeedDisable_hasNoImpact() throws Exception {
     }
 
+    @Test
+    public void userLoginAfterDisalbingFIPS() throws Exception {
```
```suggestion
    public void userLoginAfterDisablingFIPS() throws Exception {
```

jtnord committed:
fix typos

Co-authored-by: Daniel Beck <1831569+daniel-beck@users.noreply.github.com>

jtnord commented:
failure on windows appears to be the known flake that @jglick and @Vlatombe are tackling in https://github.com/jenkinsci/jenkins/pull/8534

jtnord committed:
Merge branch 'master' into HudsonPrivateSecurityRealmImplementation

MarkEWaite commented:
Congratulations on the merged pull request @divyasivasamy.  Much appreciated!

In future pull requests, it would be best if you retained the formatting from the pull request template so that the automatic changelog generation process works as expected.  I've inserted the headings into the pull request description and will confirm that the automatic changelog generation accepts the pull request, but it would be nice in future pull requests if automatic changelog maintenance could do the work instead of a person needing to do the work.

daniel-beck commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -925,34 +942,131 @@ public boolean isHashValid(String hash) {
+    /* package */ static final PasswordHashEncoder PASSWORD_HASH_ENCODER =  FIPS140.useCompliantAlgorithms() ? new PBKDF2PasswordEncoder() : new JBCryptEncoder();
+
+
+    private static final String PBKDF2 = "$PBKDF2";
+    private static final String JBCRYPT = "#jbcrypt:";
```
What motivated this inconsistency between the header formats? Asking because that seems to have permanently cut us off from using `DelegatingPasswordEncoder`.
